### PR TITLE
[WIP] Add theme scaling for graph nodes.

### PR DESCRIPTION
**Pull request by fire**
Add zoom levels instead of multiplying. Update max zoom.

WIP for review.

### CODE REVIEW

**Comment by Chaosus** [comment]
The offset for expression nodes has been removed... I think this should be untouched
image

Also I suggest you to implement a boolean switch to old zoom scroll behavior for GraphEdit - this class used not only by the Godot editor but also by the users. Enable it for Godot shader/visual script editor by default - that will be correct and dont break up compatibility.

**Comment by fire** [comment]
That's a bug with either how the zooming works or the way that some visual shaders don't apply the theme's values. Like for instance hard coding 28 as a number.

Reduz told me to do it this way so that changes for the 4-6 things that use the graph edit node also get this.

**Comment by Chaosus** [comment]
> iFire  07:51:14
> chaosus89: I'm trying to get the buttons zoomed for the pr got any ideas?
> when zoomed out the letters are too big for the button

Ah, that will fix that

```
 add_input_btn->set_h_size_flags(SIZE_EXPAND_FILL);
 add_output_btn->set_h_size_flags(SIZE_EXPAND_FILL);
```

**Comment by Chaosus** [comment]
> That's a bug with either how the zooming works or the way that some visual shaders don't apply the theme's values. Like for instance hard coding 28 as a number.

Maybe its make sense to try add an empty control with that minimum height after the expression box..

> Reduz told me to do it this way so that changes for the 4-6 things that use the graph edit node also get this.

Ok, then :) reduz is boss

**Comment by fire** [comment]
add_input_btn->set_h_size_flags(SIZE_EXPAND_FILL);


Worked after removing the spacer.
Screen Shot 2019-05-21 at 11 52 47 PM

**Comment by fire** [comment]
Adding two spacers looks ok.

Screen Shot 2019-05-21 at 11 57 12 PM

**Comment by Chaosus** [comment]
Also I've see that you've decreased the amount of zoom levels - only 3 of them are exist now - is it intended ?

**Comment by fire** [comment]
Related issue requests:

* Enable zoom with mousewheel in the visual shader editor #5214
* Zooming out or in makes text unreadable #16220
* Improve the zoom buttons on graph editors #16488
* Zoom visual script editor via mouse scrolling #6168

**Comment by fire** [comment]
@Chaosus Do we want a larger zoom or more steps between zooms?

**Comment by Chaosus** [comment]
@fire I think yes, 3 is way too small number for that, the user is not getting satisfaction when the wheel is scrolled.

**Comment by fire** [comment]
So:

1. a larger zoom?
2. more steps between zooms?

**Comment by Jummit** [comment]
I would love zooming like it is in the code editor or in the 2d viewport. Will this be that smooth?

**Comment by Calinou** [comment]
> Creating a GraphEdit in the 2D viewport and zooming there is very smooth. Will it be like that?

It doesn't work the same way at all. When you zoom on the 2D viewport, it doesn't re-render anything, causing stuff to become pixelated or blurry. Here, we want crisp visuals at any zoom level :slightly_smiling_face:

**Comment by ZX-WT** [comment]
Here's another related issue, leaving it here as a reminder to close if fixed.
https://github.com/godotengine/godot/issues/21686

**Comment by fire** [comment]
Adjusted the zoom levels for more possibilties.  
```
#define ZOOM_SCALE 1.075

#define MIN_ZOOM (((1 / ZOOM_SCALE) / ZOOM_SCALE) / ZOOM_SCALE)
#define MAX_ZOOM (((1 * ZOOM_SCALE) * ZOOM_SCALE) * ZOOM_SCALE)
```

File: scene/gui/graph_edit.cpp [hunk 1]
```diff
@@ -71,6 +72,23 @@ Error GraphEdit::connect_node(const StringName &p_from, int p_from_port, const S
 	return OK;
 }
 
+void GraphEdit::add_zoom_level(const float p_zoom_level) {
```
Comment (reduz) [inline scene/gui/graph_edit.cpp]: I would probably use indices to refer to zoom levels, not a float.. an API like
void add_zoom_level(float,Theme)
int get_zoom_level_count()
void remove_zoom_level(int)

And it's kind of your responsibility to add them in order, so the - and + buttons work properly

File: scene/gui/graph_edit.cpp [hunk 2]
```diff
@@ -1385,4 +1406,12 @@ GraphEdit::GraphEdit() {
 	setting_scroll_ofs = false;
 	just_disconnected = false;
 	set_clip_contents(true);
+	theme = get_theme();
```
Comment (reduz) [inline scene/gui/graph_edit.cpp]: Don't add them by default, let the user add them if needed. By default it should have no zoom levels and zoom controls should not be visible.

**Review by reduz — CHANGES_REQUESTED** [review]

File: scene/gui/graph_edit.cpp (hunk 2 above)
Comment (fire) [reply scene/gui/graph_edit.cpp]: Can you explain this in another way? The user? The game engine developer?

**Comment by fire** [comment]
I'm blocked by the inability to convert from mouse position to scroll offset.

```cpp
float zoom_0 = get_zoom();
_zoom_minus();
float zoom_1 = get_zoom();
Vector2 camera_0 = get_scroll_ofs() + get_size() / 2;
Vector2 mouse = b->get_position();
// https://www.reddit.com/r/godot/comments/4xo9zv/camera_zoom_to_mouse/d6hcnos
// k1 = (k0 - m) * z1 / z0 + m
// k0 is the original position of the camera,
// k1 is the new position of the camera,
// m is the position of the mouse (in the same coordinate space as the camera,
// not on screen or relative to the camera),
// and z0 and z1 are the old and new zoom values.
set_scroll_ofs((camera_0 - mouse) * zoom_1 / zoom_0 + mouse);
```

**Comment by fire** [comment]
I was unable to get the close button to work fully, but it looks good.

Video of the scaling

**Comment by fire** [comment]
This pull request slows the editor launch, I'm not sure how to optimize it.

**Comment by fire** [comment]
https://github.com/godotengine/godot/pull/35325 Significantly improves performance.

According to https://github.com/godotengine/godot/issues/35321, icon generation uses the most time.

**Comment by fire** [comment]
Update to Godot master. This is the vulkan master.

**Comment by fire** [comment]
This is salvage-able but concentrating on other things.

**Comment by jak6jak** [comment]
I'm curious what is required to complete this pull request. Could I possibly salvage this?

**Comment by fire** [comment]
The concept behind the work was that we generate 4 versions of the theme, then we flip between them on the fly.

**Comment by jak6jak** [comment]
What can I do to help?

**Comment by fire** [comment]
The patch is on 3.2 and I don’t think making like 8 resize themes is the correct approach. It should scale dynamically. However, not got any approaches. 

On Sep 13, 2020, at 7:47 PM, jak6jak <notifications@github.com> wrote:

﻿
What can I do to help?

**Comment by jak6jak** [comment]
I am looking through the code that you changed and I have a couple questions. What is a resize theme? When I think of themes I think of like colors. I noticed that you have a couple of zoom levels is that what you meant by resize themes? What do you mean by scale dynamically? 

Sorry For all the questions. Thanks for helping me out so far!

**Comment by fire** [comment]
We generate ~4 sets of themes. Each theme is a different size. I thought this was a dead end because it causes a large lag at the start of the editor.

Hope that helps.

**Pull request closed without merging**